File: sorting/mergeSort.js
const merge = function (leftArr, rightArr, arr) {
  let leftPtr = 0;
  let rightPtr = 0;
  let mergePtr = 0;

  while ((leftPtr < leftArr.length) & (rightPtr < rightArr.length)) {
    if (leftArr[leftPtr] < rightArr[rightPtr]) {
      arr[mergePtr] = leftArr[leftPtr];
      leftPtr++;
    } else {
      arr[mergePtr] = rightArr[rightPtr];
      rightPtr++;
    }
    mergePtr++;
  }

  while (leftPtr < leftArr.length) {
    arr[mergePtr] = leftArr[leftPtr];
    leftPtr++;
    mergePtr++;
  }

  while (rightPtr < rightArr.length) {
    arr[mergePtr] = rightArr[rightPtr];
    rightPtr++;
    mergePtr++;
  }

  return arr;
};

const mergeSort = function (arr) {
  if (arr.length <= 1) {
    return arr;
  }

  let start = 0;
  let end = arr.length - 1;
  let mid = Math.trunc((start + end) / 2);

  const leftArr = mergeSort(arr.slice(start, mid + 1));
  const rightArr = mergeSort(arr.slice(mid + 1, end + 1));

  return merge(leftArr, rightArr, arr);
};

console.log(mergeSort([0, 9, 8, 2, 6, 5, 1]));

File: sorting/mergeSort.cpp
# include <iostream>
using namespace std;

void merge(int arr[], int start, int mid, int end) {
  int leftArrSize = mid - start + 1;
  int rightArrSize = end - mid;

  int leftArr[leftArrSize];
  int rightArr[rightArrSize];

  for (int i = 0; i < leftArrSize; i++) {
    leftArr[i] = arr[i + start];
  }

  for (int i = 0; i < rightArrSize; i++) {
    rightArr[i] = arr[i + mid + 1];
  }

  int leftPtr = 0;
  int rightPtr = 0;
  int mergePtr = start;

  while (leftPtr < leftArrSize && rightPtr < rightArrSize) {
    if (leftArr[leftPtr] < rightArr[rightPtr]) {
      arr[mergePtr] = leftArr[leftPtr];
      leftPtr++;
    } else {
      arr[mergePtr] = rightArr[rightPtr]; 
      rightPtr++;
    }
    mergePtr++;
  }

  while (leftPtr < leftArrSize) {
    arr[mergePtr] = leftArr[leftPtr];
    leftPtr++;
    mergePtr++;
  }

  while (rightPtr < rightArrSize) {
    arr[mergePtr] = rightArr[rightPtr];
    rightPtr++;
    mergePtr++;
  }
}


void mergeSort(int arr[], int start, int end) {
  if (start >= end) {
    return;
  }

  int mid = start + (end - start) / 2;

  mergeSort(arr, start, mid);
  mergeSort(arr, mid + 1, end);
  merge(arr, start, mid, end);
}

int main() {
  int arr[] = {5, 3, 1, 0, 8, 6, 12, 23, 0, 9};
  int size = sizeof(arr) / sizeof(arr[0]);
for (int i = 0; i < size; i++) {
    cout << arr[i] << " ";
  }

  cout << endl;

  mergeSort(arr, 0, size - 1);
   for (int i = 0; i < size; i++) {
    cout << arr[i] << " ";
  }
  return 0;
}